### DISCUSSION TITLE
Select an app to open winget

### DISCUSSION THREAD

**Discussion by paul-8235d9** [microsoft/winget-cli#4423] [Configuration]
I have a clean install of Windows 11 but have had nothing but issues with winget. Originally, it was winget only replied; "No package found matching input criteria" I manually down loaded the msi to update from 1.2.10691 to 1.7.10861. I've now managed to "fix" the problem so now any winget command in PowerShell will open a popup saying to Select an app to open 'winget'.  In Cmd (as admin) I get 

C:\Windows\System32>winget --version
v1.7.10861
C:\Windows\System32>winget list --id Microsoft.Powershell
No installed package found matching input criteria.
C:\Windows\System32>winget list --id Powershell
No installed package found matching input criteria.
C:\Windows\System32>winget install Microsoft.PowerToys --source winget
No sources match the given value: winget
The configured sources are:
  msstore
![winget Screenshot 2024-04-29 100229](https://github.com/microsoft/winget-cli/assets/25016924/45e20462-d991-4d1b-bcea-6784cb0dac6f)

**Comment by Trenly** [comment]
Try running this command `Add-AppxPackage https://aka.ms/getwinget` and see if that does anything to fix the issue, if it doesn't you may need to use `Get-AppxPackage 'Microsoft.DesktopAppInstaller' | Remove-AppxPackage; Add-AppxPackage https://aka.ms/getwinget`

**Reply by paul-8235d9** [reply]
The Add-AppxPackage command looked promising, I mean it looked like it was processing and doing stuff. But I still get the Select an app to open winget popup.

The second command failed (you can see in the image, I am running PS as administrator):

Remove-AppxPackage: Deployment failed with HRESULT: 0x80073CFA, Removal failed. Please contact your software vendor. (Exception from HRESULT: 0x80073CFA)
error 0x80070032: AppX Deployment Remove operation on package Microsoft.DesktopAppInstaller_1.22.11132.0_x64__8wekyb3d8bbwe from: C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.22.11132.0_x64__8wekyb3d8bbwe failed. This app is part of Windows and cannot be uninstalled on a per-user basis. An administrator can attempt to remove the app from the computer using Turn Windows Features on or off. However, it may not be possible to uninstall the app.

NOTE: For additional information, look for [ActivityId] d08d6bc7-9b2e-0000-e7f1-8fd02e9bda01 in the Event Log or use the command line Get-AppPackageLog -ActivityID d08d6bc7-9b2e-0000-e7f1-8fd02e9bda01

![winget Screenshot 2024-04-30 123345](https://github.com/microsoft/winget-cli/assets/25016924/f85b8d92-5c62-426a-a8fa-caf58e552154)

Running the Get-AppPackageLog command gave zero information.
![winget Screenshot 2024-04-30 123726](https://github.com/microsoft/winget-cli/assets/25016924/0c65b2f2-1084-49ed-8833-d518ed7ffc2d)

**Reply by Trenly** [reply]
Hmm. Quite odd. Usually a removal then a re-install works for me. Have you checked the app execution alias is set correctly?

https://github.com/microsoft/winget-cli/tree/master/doc/troubleshooting#executing-winget-doesnt-display-help

**Reply by paul-8235d9** [reply]
### Powershell 7 vs. CMD
* Both run as elevated user
* Running `winget --version`
	* On PowerShell get a popup saying to Select an app to open 'winget'
	* ON CMD returns: v1.7.11132
## Updating winget
I wanted to see if updating winget would fix the issue.
### Via Store
Install the latest version of WinGet through the Microsoft Store by installing the latest version of [App Installer](https://apps.microsoft.com/detail/9NBLGGH4NNS1)
- The link opens a webpage for App Installer but doesn't show what version it is But the release date is 4/15/2016 which seems odd to say the least. Clicking on **Install** launches the Microsoft Store App.
- In the App under additional information it says version 1.22.1132.0, released 4/15/2016
- Running `winget --version` in both PS and CMD gives the same results as before. Note - Version is 1.7.11132. If the App store is correct, and I doubt it because of the release date, that winget might not have been updated.
### Via GitHub release
- GitHub says **Windows Package Manager 1.7.11132**, which is what is reported on my system.
- Downloaded `Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle`
- Clicked on the downloaded file, it reports it is already installed. Selected **Reinstall** 
	- Note another bug; I get a popup saying a newer version is already installed; You already have a newer version 1.22.11132.0 of this installation installed are you sure you want to reinstall version 1.22.11132.0?
		- I downloaded Package Manager 1.7.11132, not 1.22.1132.0
		- Since when is 1.22.11132.0 != 1.22.11132.0 ?
- Both PS and CMD exhibit the same behavior. Note - selecting **Reinstall** seemed to happen very quickly. Did it do anything?
## Executing `winget` doesn't display help
This is true in PS, not in CMD. But what the hell ...
1. Executing `Get-AppxPackage microsoft.desktopappinstaller`
``` PowerShell
RunspaceId             : c2d68e9f-8f20-4972-84f6-d42d3111e891
Name                   : Microsoft.DesktopAppInstaller
Publisher              : CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
PublisherId            : 8wekyb3d8bbwe
Architecture           : X64
ResourceId             :
Version                : 1.22.11132.0
PackageFamilyName      : Microsoft.DesktopAppInstaller_8wekyb3d8bbwe
PackageFullName        : Microsoft.DesktopAppInstaller_1.22.11132.0_x64__8wekyb3d8bbwe
InstallLocation        : C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.22.11132.0_x64__8wekyb3d8bbwe
IsFramework            : False
PackageUserInformation : {}
IsResourcePackage      : False
IsBundle               : False
IsDevelopmentMode      : False
NonRemovable           : True
Dependencies           : {Microsoft.UI.Xaml.2.8_8.2310.30001.0_x64__8wekyb3d8bbwe,
                         Microsoft.VCLibs.140.00.UWPDesktop_14.0.33519.0_x64__8wekyb3d8bbwe,
                         Microsoft.DesktopAppInstaller_1.22.11132.0_neutral_split.scale-100_8wekyb3d8bbwe}
IsPartiallyStaged      : False
SignatureKind          : Store
Status                 : Ok
```
2. Go to `App execution aliases` option in `Apps & features Settings` to enable it
	- In Settings > Apps > Advanced app settings > App execution aliases > Windows Package Manager Client (winget.exe) is selected On.
3. The App Installer did not automatically add the PATH environment variable
	- In CMD `%LOCALAPPDATA%\Microsoft\WindowsApps\winget` returned the winget help page.
	- IN PS `& "$env:LOCALAPPDATA\Microsoft\WindowsApps\winget"` ALSO returns the help page!
	- My path is:
	```
	C:\Users\paul>path
	PATH=C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\Program Files\PowerShell\7\;C:\Program Files\Tailscale\;C:\Program Files\Git\cmd;C:\Users\paul\AppData\Local\Microsoft\WindowsApps;C:\Users\paul\.dotnet\tools;C:\Users\paul\AppData\Local\Programs\Microsoft VS Code\bin
	```
### So, something seems wrong with PS and or my understanding of Path
I believe `C:\Users\paul\AppData\Local\Microsoft\WindowsApps` == `%LOCALAPPDATA%\Microsoft\WindowsApps` So, why doesn't `winget --help` work in PS?

EDIT to add

- I checked the Environment Variables (Windows button > environment variables > Advanced > Environment Variables) and noticed:
	- Under User variables the path includes `%USERPROFILE%\AppData\Local\Microsoft\WindowsApps`
	- Under System variables it is not included in either Path or PSModulePath

**Comment by davidldennison** [comment]
Not sure of the answer, but I've been tweaking and breaking Winget for a bit now I had this exact issue and managed to solve it, but can't remember how - but I think it's 1, 2, & 3. Here are some additional things to add to such an already exhaustive list of documenting your troubleshooting steps (great job btw!): 

### 1. Try putting all those directories on your SYSTEM and USER path.

### 2. Add ```C:\Program Files\WindowsApps\``` & "C:\Program Files\WinGet" as well.

### 3. I've also noticed _weird_ stuff happening that I solved by adding absolute path AND the "Expandable String" versions of those directories on your path:
 ```%USERPROFILE%\AppData\Local\Microsoft\WindowsApps```
```C:\Users\david\AppData\Local\Microsoft\WindowsApps```

Certain installations, installation methods, and apps seem to sometimes specifically be looking for one or the other and if it doesn't get the expected path, it could cause problems.

### 5. Try installing winget and app installer packages with MSIX Hero

This is also a GREAT debugging tool that lets you dive deep into apps as well as lets you see if you are missing dependencies for installed apps.

Get it here: https://github.com/marcinotorowski/MSIX-Hero

### 4. This is way more involved but install the latest release of the MSIX SDK here: https://github.com/microsoft/msix-packaging/releases

I *believe* the SDK registers itself as an app-installer and as such next time you go to try to install an MSIX, it will give you the option to select an installer as a sort of "reset" with two options - one being the App-Installer and the other being the SDK.

Go back to terminal and repeat your winget command. You should see both options - select App-Installer. 

If you don't see app installer, uninstall both App-Installer AND Winget. Clear everything. Download the .msixbundle again from the releases page (make sure it's unblocked) and install and register directly with DISM commands directly in CMD as admin. (A great tool if you need help with visualizing - also great for debugging - DISM: https://github.com/CodingWonders/DISMTools).

One or both of these commands *should* give you a sort of App-Installer reset that also resets App execution alias AT THE SAME TIME - which seems to be important.

### 5. The last thing I can think of is your PowerShell environment, directories, and path.

Between PowerShell using different folders, those folders needing to be on the path, those needing to be in ```PSModulePath```, what version of PS and modules you have, AND ultimately how PS parses the PATH in different user contexts it can get messy. I could write another 20 bullet points for this, but that is a little out of scope for this post. But Rapid Environment Editor has been a huge help: 

https://www.rapidee.com/en/download

Let me know if that helps!

**Reply by paul-8235d9** [reply]
I'm not ignoring  your suggestions. I'm doing a backpack trip up in the Mtns and just caught a cell signal and sat down and read some of my messages. I'll give your suggestion a try when I'm home.

**Reply by davidldennison** [reply]
Hope you had a good trip! Any chance you were able to try some of these? Would love to see if you made any discoveries on this issue!
